nncf/experimental/torch/sparsify_activations/torch_backend.py

```diff
@@ -0,0 +1,194 @@
+# Copyright (c) 2024 Intel Corporation
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#      http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Dict, List, Type, TypeVar
+
+import numpy as np
+import torch
+import torch.nn as nn
+
+import nncf
+from nncf.common.graph.graph import NNCFGraph
+from nncf.common.graph.graph import NNCFNode
+from nncf.common.graph.operator_metatypes import CONST_NOOP_METATYPES
+from nncf.common.graph.operator_metatypes import OperatorMetatype
+from nncf.common.graph.transformations.commands import TargetType
+from nncf.common.graph.transformations.layout import TransformationLayout
+from nncf.data import Dataset
+from nncf.experimental.torch.sparsify_activations.sparsify_activations_impl import SparsifyActivationsAlgoBackend
+from nncf.torch.graph import operator_metatypes as om
+from nncf.torch.graph.transformations.commands import PTSharedFnInsertionCommand
+from nncf.torch.graph.transformations.commands import PTTargetPoint
+from nncf.torch.model_transformer import PTModelTransformer
+from nncf.torch.nncf_network import NNCFNetwork
+from nncf.torch.utils import training_mode_switcher
+
+ACTIVATIONS_SPARSIFIER_PREFIX = "activations_sparsifier"
+TModel = TypeVar("TModel")
+
+
+class ActivationsSparsifier(nn.Module):
+    """
+    Sparsifies input activations by masking out values around zero.
+    """
+
+    def __init__(self, target_sparsity: float, alpha: float = 0.2):
+        """
+        :param target_sparsity: The target activation sparsity level.
+        :param alpha: The exponential moving average decay factor in range (0, 1) for calibrating
+            the threshold. A larger alpha will give more weight to the most recent batches.
+        """
+        super().__init__()
+        self.target_sparsity = target_sparsity
+        if alpha <= 0.0 or alpha >= 1.0:
+            raise ValueError("The decay factor `alpha` should be in range (0, 1).")
+        self.alpha = alpha
+        self.register_buffer("running_threshold", torch.tensor(0.0))
+        self.register_buffer("num_batches_tracked", torch.tensor(0))
+        self.running_threshold: torch.Tensor
+        self.num_batches_tracked: torch.Tensor
+        self._freeze = False
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        if not self._freeze:
+            threshold = self._calculate_threshold(x, self.target_sparsity)
+            self._update(threshold)
+        mask = torch.le(x.abs(), self.running_threshold)
+        x = torch.masked_fill(x, mask, 0.0)
```

Comment:
I don't think Inplace is a good idea here. It will affect other layers that use the same input activations. E.g., in Llama, both up_proj and gate_proj receives the same input x, and let's assume we want to sparsify up_proj only.

Reply:
I see, thanks